Documentation for replacing the Old Chado Organism API methods with new OOP methods (#142)
tripal/tripal_doc: docs/dev_guide/deprecations/chado_organism_api.rst

Tripal Chado Organism API is deprecated in favour of the ChadoOrganismBuddy and ChadoOrganismFormElementController
===================================================================================================================

- **Deprecated in** tripal 4.0.0-alpha4
- **Removed in** tripal 4.1.0
- **Issue** `#2424 <https://github.com/tripal/tripal/issues/2424>`_
- **PR** `#2426 <https://github.com/tripal/tripal/pull/2426>`_

> Methods in ChadoOrganismBuddy class and ChadoOrganismFormElementController class replace the functionality of the existing Chado Organism API methods.

Reviewer: There is a good chance that someone who previously used the Chado Organism API doesn't know what Chado Buddies are yet, so we should give them some guidance on how to use them. How about:

```suggestion
Methods in ChadoOrganismBuddy class and ChadoOrganismFormElementController class replace the majority of the functionality of the existing Chado Organism API methods. 

Some of these methods can be provided a ChadyBuddyRecord as a parameter, and may return one or more Organism buddies. To learn more about ChadoBuddies, refer to: [link to Chado Buddies section of the docs]
```

> Methods in ChadoOrganismBuddy class and ChadoOrganismFormElementController class replace the functionality of the existing Chado Organism API methods.

Reviewer 2: ```suggestion
Methods in ChadoOrganismBuddy and ChadoOrganismFormElementController classes replace the functionality of the existing Chado Organism API methods.
```

> 

Reviewer 2: Should we add a line that states something like? - The examples shown here use static service calls. When integrating services, prefer Dependency Injection as recommended by Drupal best practices.

Reviewer: I agree 👍 Although maybe this can be placed into its own "Note" tag. You can also link directly to instructions for injecting the ChadoBuddy service: https://tripaldoc.readthedocs.io/en/latest/dev_guide/biodata/buddies/inject.html#injecting-the-buddy-service

chado_get_organism()
--------------------

**Before:**

.. code-block:: php

    $identifiers = [
        'genus' => 'Lens',

>         'species' => 'culinaris',

Reviewer: Because these are Tripal docs, I think we should stick to more generic examples like "Tripalus databasica". 
```suggestion
        'genus' => 'Tripalus',
        'species' => 'databasica',
```

    ];

>     $organism = chado_get_organism($identifiers, []);

Reviewer: ```suggestion
    $organism_object = chado_get_organism($identifiers, []);
```

**After:**

.. code-block:: php

    $buddy_service = \Drupal::service('tripal_chado.chado_buddy');
    $organism_buddy_instance = $buddy_service->createInstance('chado_organism_buddy', []);

    $conditions = [
        'organism.genus' => 'Lens',

>         'organism.species' => 'culinaris',

Reviewer: ```suggestion
        'organism.genus' => 'Tripalus',
        'organism.species' => 'databasica',
```

    ];

>     $organism = $organism_buddy_instance->getOrganism($conditions, []);

Reviewer: ```suggestion
    // Retrieves an array of ChadoOrganismBuddy records. This array will be empty if no records matched the input values, and can have one or more records if matches were found.
    $organism_buddies = $organism_buddy_instance->getOrganism($conditions, []);
```

chado_get_organism_scientific_name()
------------------------------------

**Before:**

.. code-block:: php

>     $organism = chado_get_organism_scientific_name($organism_object);

Reviewer: ```suggestion
    // Returns a string of the matching organism's scientific name
    $organism_name = chado_get_organism_scientific_name($organism_object);
```

**After:**

.. code-block:: php

    $buddy_service = \Drupal::service('tripal_chado.chado_buddy');
    $organism_buddy_instance = $buddy_service->createInstance('chado_organism_buddy', []);

>     $organism = $organism_buddy_instance->getOrganismScientificName($organism_object, []);

Reviewer: ```suggestion
    $conditions = [
        'organism.common_name' => 'Tripal',
    ];

    // Returns a string of the matching organism's scientific name
    $organism_name = $organism_buddy_instance->getOrganismScientificName($conditions);
```

chado_get_organism_select_options()
-----------------------------------

**Before:**

.. code-block:: php

    $organisms = chado_get_organism_select_options();

**After:**

.. code-block:: php

  use Drupal\tripal_chado\Controller\ChadoOrganismFormElementController;

  ...

  $organisms = ChadoOrganismFormElementController::getSelectOptions([]);

chado_get_organism_id_from_scientific_name()
--------------------------------------------

**Before:**

.. code-block:: php

>     $organism_id = chado_get_organism_id_from_scientific_name($scientific_name, $options);

Reviewer: I think adding another example here for common name helps to demonstrate that the options between these methods have remained the same!
```suggestion
    By Scientific Name:
    $scientific_name = 'Tripalus databasica';
    // Returns an array of organism IDs that match the specified scientific name.
    $organism_ids = chado_get_organism_id_from_scientific_name($scientific_name);
    
    By Common Name:
    $common_name = 'Tripal';
    // Returns an array of organism IDs that match the specified common name.
    $organism_ids = chado_get_organism_id_from_scientific_name($common_name, ['check_common_name' => TRUE]);
```

**After:**

.. code-block:: php

    $buddy_service = \Drupal::service('tripal_chado.chado_buddy');
    $organism_buddy_instance = $buddy_service->createInstance('chado_organism_buddy', []);

>     $organism_id = $organism_buddy_instance->getOrganismFromScientificName($scientific_name, $options);

Reviewer: ```suggestion
    By Scientific Name:
    $scientific_name = 'Tripalus databasica';
     // Retrieves an array of ChadoOrganismBuddy records. This array will be empty if no records matched the input values, and can have one or more records if matches were found.
    $organism_buddies = $organism_buddy_instance->getOrganismFromScientificName($scientific_name);
    // Then grab the organism's ID
    $organism_id = $organism_buddies[0]->getValue('organism.organism_id');
    
    By Common Name:
    $common_name = 'Tripal';
    $organism_buddies = $organism_buddy_instance->getOrganismFromScientificName($common_name, ['check_common_name' => TRUE]);
    // Then grab the organism's ID
    $organism_id = $organism_buddies[0]->getValue('organism.organism_id');
```

chado_abbreviate_infraspecific_rank()
-------------------------------------

**Before:**

.. code-block:: php

>     $abbreviation = chado_abbreviate_infraspecific_rank($rank_species);

Reviewer: ```suggestion
    // Returns the abbreviation of an organism's rank.
    $rank_abbreviation = chado_abbreviate_infraspecific_rank($rank);
```

**After:**

.. code-block:: php

    $buddy_service = \Drupal::service('tripal_chado.chado_buddy');
    $organism_buddy_instance = $buddy_service->createInstance('chado_organism_buddy', []);

>     $abbreviation = $organism_buddy_instance->abbreviateInfraspecificRank($rank_species);

Reviewer: ```suggestion
    $rank_abbreviation = $organism_buddy_instance->abbreviateInfraspecificRank($rank);
```

chado_unabbreviate_infraspecific_rank()
----------------------------------------

**Before:**

.. code-block:: php

>     $rank = chado_unabbreviate_infraspecific_rank($rank_species);

Reviewer: ```suggestion
    $rank = chado_unabbreviate_infraspecific_rank($rank_abbreviation);
```

**After:**

.. code-block:: php

    $buddy_service = \Drupal::service('tripal_chado.chado_buddy');
    $organism_buddy_instance = $buddy_service->createInstance('chado_organism_buddy', []);

>     $rank = $organism_buddy_instance->unabbreviateInfraspecificRank($rank_species);

Reviewer: ```suggestion
    $rank = $organism_buddy_instance->unabbreviateInfraspecificRank($rank_abbreviation);
```

Tripal Chado Organism API is deprecated in favour of the ChadoOrganismBuddy and ChadoOrganismFormElementController
===================================================================================================================

- **Deprecated in** tripal 4.0.0-alpha4
- **Removed in** tripal 4.1.0
- **Issue** `#2424 <https://github.com/tripal/tripal/issues/2424>`_
- **PR** `#2426 <https://github.com/tripal/tripal/pull/2426>`_

Methods in ChadoOrganismBuddy and ChadoOrganismFormElementController classes replace the majority of the functionality of the existing Chado Organism API methods. 

Some of these methods can be provided a ChadyBuddyRecord as a parameter, and may return one or more Organism buddies. To learn more about ChadoBuddies, refer to: `Chado Buddies Documentation <https://tripaldoc.readthedocs.io/en/latest/dev_guide/biodata/buddies.html>`_

.. note::

>     The examples below demonstrate static service calls to the ChadoOrganismBuddy and ChadoOrganismFormElementController classes; however, Dependency Injection is preferred when integrating services, as recommended by Drupal best practices. For more details regarding Injecting ChadoBuddy Services, refer to: `Injecting the Buddy Service Documentation <https://tripaldoc.readthedocs.io/en/latest/dev_guide/biodata/buddies/inject.html#injecting-the-buddy-service>`_

Reviewer 3: ```suggestion
Some of these methods can be provided a ChadyBuddyRecord as a parameter, and may return one or more Organism buddies. To learn more about ChadoBuddies, refer to: :ref:`Chado Buddies Documentation <Chado Buddies>`

.. note::
    The examples below demonstrate static service calls to the ChadoOrganismBuddy and ChadoOrganismFormElementController classes; however, Dependency Injection is preferred when integrating services, as recommended by Drupal best practices. For more details regarding Injecting ChadoBuddy Services, refer to: :ref:`Injecting the Buddy Service Documentation <Injecting the Buddy Service>`
```
Better style is to use :ref: rather than hardcoded url, and then you can test build locally too.

chado_get_organism()
--------------------

**Before:**

.. code-block:: php

    $identifiers = [
        'genus' => 'Tripalus',  
        'species' => 'databasica',
    ];
    $organism_object = chado_get_organism($identifiers, []);

**After:**

.. code-block:: php

    $buddy_service = \Drupal::service('tripal_chado.chado_buddy');
    $organism_buddy_instance = $buddy_service->createInstance('chado_organism_buddy', []);

    $conditions = [
        'organism.genus' => 'Tripalus',  
        'organism.species' => 'databasica',
    ];

>     // Retrieves an array of ChadoOrganismBuddy records. This array will be empty if no records matched the input values, and can have one or more records if matches were found.  

Reviewer 3: ```suggestion
    // Retrieves an array of ChadoOrganismBuddy records. This array will be empty if no records matched
    // the input values, and can have one or more records if matches were found.  
```
Wrap lines so we don't see horizontal scroll bars when viewing these examples.
<img width="704" height="200" alt="Image" src="https://github.com/user-attachments/assets/1939f637-91c2-4eb8-a917-7718215112ba" />

chado_get_organism_id_from_scientific_name()
--------------------------------------------

**Before:**

By Scientific Name:

.. code-block:: php

    $scientific_name = 'Tripalus databasica';  
    // Returns an array of organism IDs that match the specified scientific name.  
    $organism_ids = chado_get_organism_id_from_scientific_name($scientific_name);  
    
By Common Name:

.. code-block:: php

    $common_name = 'Tripal';  
    // Returns an array of organism IDs that match the specified common name.  
    $organism_ids = chado_get_organism_id_from_scientific_name($common_name, ['check_common_name' => TRUE]);

**After:**

By Scientific Name:  

.. code-block:: php

    $scientific_name = 'Tripalus databasica';  
    // Retrieves an array of ChadoOrganismBuddy records. This array will be empty if no records matched the input values, and can have one or more records if matches were found.  
    $organism_buddies = $organism_buddy_instance->getOrganismFromScientificName($scientific_name);  
    // Then grab the organism's ID  
    $organism_id = $organism_buddies[0]->getValue('organism.organism_id');  

By Common Name:

.. code-block:: php

    $common_name = 'Tripal';  

>     $organism_buddies = $organism_buddy_instance->getOrganismFromScientificName($common_name, ['check_common_name' => TRUE]);  

Reviewer 3: ```suggestion
    $organism_ids = chado_get_organism_id_from_scientific_name($common_name,
      ['check_common_name' => TRUE]);

**After:**

By Scientific Name:  

.. code-block:: php

    $scientific_name = 'Tripalus databasica';  
    // Retrieves an array of ChadoOrganismBuddy records. This array will be empty if no records matched
    // the input values, and can have one or more records if matches were found.  
    $organism_buddies = $organism_buddy_instance->getOrganismFromScientificName($scientific_name);  
    // Then grab the organism's ID  
    $organism_id = $organism_buddies[0]->getValue('organism.organism_id');  

By Common Name:

.. code-block:: php

    $common_name = 'Tripal';  
    $organism_buddies = $organism_buddy_instance->getOrganismFromScientificName($common_name,
      ['check_common_name' => TRUE]);  
```
Wrap lines so we don't see horizontal scroll bars when viewing these examples.
